packages/dds/tree/docs/wip/revertible-fork-api.md

```diff
@@ -0,0 +1,207 @@
+
+### Revertibles
+
+This is solved by exposing handles to commits which can (and should) be disposed when they are no longer possibly going to be reverted.
+Since these handles are specifically used by the revert functionality, we call them `Revertibles`, and they have (something like) the following interface:
+
+```ts
+interface Revertible {
+	revert();
+	dispose();
+}
+```
+
+Typically, revertibles are placed in an ordered undo and redo stack and managed by the user (perhaps via an undo/redo helper library).
+
+> Revertibles need not be reverted in order.
+> An undo/redo stack is a common use of revertibles, but other "undo-ish" data structures might be appropriate for an application as well. 
+
+Importantly, revertibles are only allocated if the user requests them to be.
+When a new commit is applied to the tree, an event (`"commitApplied"`) fires which provides the user a _function_ that they can call _if_ they want to allocate a revertible.
+If they do this, then they are also responsible for disposing the revertible as soon as they no longer need it (for example, if it ends up too far back in their undo stack).
+In this way, revertibles are given the minimal lifetime possible (if they are given a life at all) in order to reduce the overhead of maintaining them.
+
+## Branches
+
+How do revertibles interact with SharedTree [branches](../main/indexes-and-branches.md)?
+Note that a `Revertible` has a parameterless `revert()` method.
+Under the hood it is performing the fundamental `revert(commit, branch)` operation and therefore the revertible must be implicitly associated with both a commit _and_ a branch. This means that different branches will have different revertible objects, even for reverting the same commit.
+
+This is accomplished by allowing revertibles to be _cloned_ (or "forked"). Cloning a revertible produces a new revertible which is associated with the provided branch - so, when `revert()` is called on that new revertible, it applies the reverted commit to the head of that branch.
+
+> The original "non-cloned" revertible, which came from the `"commitApplied"` event, is implicitly associated with the branch that raised the event.
+
+There is a particular scenario in which this is necessary for a user.
+If a user wants to create a new branch and allow that branch to undo past the commit at which it was created (i.e., undo "back into its parent branch"), then all the relevant revertibles from the parent branch must be cloned and the clones must be associated with the new branch. For example:
+
+1. Parent branch **P** appends commits _X_, _Y_, and _Z_.
+2. Child branch **C** branches off of the current head (_Z_) of **P**.
+3. The user wants to undo _Z_ and then _Y_, but on branch **C**, not on branch **P**.
+   This would leave **P** untouched and leave **C** with a state including only the commit _X_.
```

Comment:
This sentence is confusing because it mixes up the commits that are on a branch and the state of the branch.
```suggestion
   This would leave **P** untouched and add _UndoOf(Z)_, _UndoOf(Y)_ onto **C**.
```